File: lib/src/tensor.dart
import 'dart:ffi' as ffi;
import 'dart:typed_data';

import 'package:ffi/ffi.dart';
import 'package:mnn/src/stb_image/enum.dart';
import 'package:mnn/src/stb_image/image.dart';

import 'base.dart';
import 'g/mnn.g.dart' as c;
import 'halide_runtime.dart';

class Tensor extends NativeObject {
  static final ffi.NativeFinalizer _finalizer = ffi.NativeFinalizer(c.addresses.mnn_tensor_destroy);

  Tensor.fromPointer(super.ptr, {super.attach, super.externalSize});

  factory Tensor.create({int dimSize = 4, c.DimensionType dimType = c.DimensionType.MNN_CAFFE}) {
    final p = c.mnn_tensor_create(dimSize, dimType);
    return Tensor.fromPointer(p);
  }

  factory Tensor.createDevice(
    List<int> shape,
    HalideType type, {
    c.DimensionType dimType = c.DimensionType.MNN_TENSORFLOW,
  }) {
    final pShape = calloc<ffi.Int>(shape.length);
    pShape.cast<ffi.Int32>().asTypedList(shape.length).setAll(0, shape);
    final p = c.mnn_tensor_create_device(pShape, shape.length, type.ref, dimType);
    return Tensor.fromPointer(p);
  }

  factory Tensor.fromTensor(
    Tensor tensor, {
    c.DimensionType dimType = c.DimensionType.MNN_CAFFE,
    bool allocMemory = true,
  }) {
    final p = c.mnn_tensor_create_from_tensor(tensor.ptr, dimType, allocMemory);
    return Tensor.fromPointer(p);
  }

  factory Tensor.fromData(
    List<int> shape,
    HalideType type, {
    Uint8List? data,
    c.DimensionType dimType = c.DimensionType.MNN_TENSORFLOW,
  }) {
    final pShape = calloc<ffi.Int>(shape.length);
    pShape.cast<ffi.Int32>().asTypedList(shape.length).setAll(0, shape);
    final pData = data == null ? ffi.nullptr : calloc<ffi.Uint8>(data.length);
    if (data != null) {
      pData.asTypedList(data.length).setAll(0, data);
    }
    final p = c.mnn_tensor_create_with_data(pShape, shape.length, type.ref, pData.cast(), dimType);
    return Tensor.fromPointer(p);
  }

  factory Tensor.fromImage(
    Image image, {
    c.DimensionType dimType = c.DimensionType.MNN_CAFFE,
  }) {
    final imLength = image.width * image.height * image.channels;
    final pShape = calloc<ffi.Int>(4);
    pShape.cast<ffi.Int32>().asTypedList(4).setAll(0, [1, image.channels, image.width, image.height]);
    if (image.dtype == StbiDType.f32) {
      final p = c.mnn_tensor_create_with_data(pShape, 4, HalideType.f32().ref, image.ptr, dimType);
      return Tensor.fromPointer(p, attach: false);
    } else if (image.dtype == StbiDType.u8 || image.dtype == StbiDType.u16) {
      final pData = malloc<ffi.Float>(imLength);
      for (var i = 0; i < imLength; i++) {
        pData[i] = image.values[i].toDouble();
      }
      final p = c.mnn_tensor_create_with_data(pShape, 4, HalideType.f32().ref, pData.cast(), dimType);
      return Tensor.fromPointer(p, attach: true);
    } else {
      throw Exception('Unsupported image dtype: ${image.dtype}');
    }
  }

  @override
  ffi.NativeFinalizer get finalizer => _finalizer;

  @override
  void release() {
    c.mnn_tensor_destroy(ptr);
  }

  Tensor clone({bool deepCopy = false}) {
    final p = c.mnn_tensor_clone(ptr, deepCopy);
    return Tensor.fromPointer(p);
  }

  /// @brief for DEVICE tensor, copy data from given host tensor.
  ///
  /// @param hostTensor    host tensor, the data provider.
  ///
  /// @return true for DEVICE tensor, and false for HOST tensor.
  bool copyFromHost(Tensor hostTensor) {
    final code = c.mnn_tensor_copy_from_host(ptr, hostTensor.ptr);
    return switch (code) {
      c.ErrorCode.BOOL_TRUE => true,
      c.ErrorCode.BOOL_FALSE => false,
      _ => throw Exception('copyFromHost failed: $code'),
    };
  }

  /// @brief for DEVICE tensor, copy data to given host tensor.
  ///
  /// @param hostTensor    host tensor, the data consumer.
  ///
  /// @return true for DEVICE tensor, and false for HOST tensor.
  bool copyToHost(Tensor hostTensor) {
    final code = c.mnn_tensor_copy_to_host(ptr, hostTensor.ptr);
    return switch (code) {
      c.ErrorCode.BOOL_TRUE => true,
      c.ErrorCode.BOOL_FALSE => false,
      _ => throw Exception('copyToHost failed: $code'),
    };
  }

  int get dimensions => c.mnn_tensor_dimensions(ptr);

  List<int> get shape {
    final dims = dimensions;
    final shape = calloc<ffi.Int>(dims);
    final code = c.mnn_tensor_shape(ptr, shape, dims);
    if (code != c.ErrorCode.NO_ERROR) {
      throw Exception('mnn_tensor_shape failed: $code');
    }
    final result = List<int>.generate(dims, (i) => shape[i]);
    calloc.free(shape);
    return result;
  }

  int get count => shape.reduce((value, element) => value * element);

  /// @brief Get tensor data size in bytes
  ///
  /// @return Size in bytes
  int get size => c.mnn_tensor_size(ptr);

  /// @brief Get tensor element count
  ///
  /// @return Element count
  int get elementSize => c.mnn_tensor_element_size(ptr);

  /// @brief Get tensor shape in bytes (unsigned)
  ///
  /// @return Size in bytes
  int get usize => c.mnn_tensor_usize(ptr);

  /// @brief Get tensor width
  ///
  /// @return Width
  int get width => c.mnn_tensor_width(ptr);

  /// @brief Get tensor height
  ///
  /// @return Height
  int get height => c.mnn_tensor_height(ptr);

  /// @brief Get tensor channel
  ///
  /// @return Channel
  int get channel => c.mnn_tensor_channel(ptr);

  /// @brief Get tensor batch
  ///
  /// @return Batch
  int get batch => c.mnn_tensor_batch(ptr);

  /// @brief Get tensor stride
  ///
  /// @param [index] Dimension index
  ///
  /// @return Stride
  int getStride(int index) => c.mnn_tensor_stride(ptr, index);

  /// @brief Get tensor length
  ///
  /// @param [index] Dimension index
  ///
  /// @return Length
  int getLength(int index) => c.mnn_tensor_length(ptr, index);

  /// @brief Set tensor stride
  ///
  /// @param [index] Dimension index
  ///
  /// @param [stride] Stride value
  void setStride(int index, int stride) => c.mnn_tensor_set_stride(ptr, index, stride);

  /// @brief Set tensor length
  ///
  /// @param index Dimension index
  ///
  /// @param length Length value
  void setLength(int index, int length) => c.mnn_tensor_set_length(ptr, index, length);

  /// @brief Get host data pointer
  ///
  /// @return Data pointer or NULL
  ffi.Pointer<ffi.Void> get host => c.mnn_tensor_host(ptr);

  /// @brief Get device ID
  ///
  /// @return Device ID
  int get deviceId => c.mnn_tensor_device_id(ptr);

  /// @brief Get buffer
  ///
  /// @return Buffer pointer
  HalideBuffer get buffer => HalideBuffer.fromPointer(c.mnn_tensor_buffer(ptr), attach: false);

  c.DimensionType get dimensionType => c.mnn_tensor_get_dimension_type(ptr);
  c.HandleDataType get handleDataType => c.mnn_tensor_get_handle_data_type(ptr);

  /// @brief Set data type
  ///
  /// @param type Data type
  void setDataType(DataType type) => c.mnn_tensor_set_type(ptr, type.value);

  /// @brief Get data type
  ///
  /// @return Data type
  HalideType get type => HalideType.fromPointer(c.mnn_tensor_get_type(ptr));

  /// @brief Map tensor for access
  ///
  /// @param mtype Map type
  ///
  /// @param dtype Dimension type
  ///
  /// @return Mapped pointer or NULL
  ffi.Pointer<ffi.Void> map(c.MapType mtype, c.DimensionType dtype) => c.mnn_tensor_map(ptr, mtype, dtype);

  /// @brief Unmap tensor
  ///
  /// @param mtype Map type
  ///
  /// @param dtype Dimension type
  ///
  /// @param map_ptr Mapped pointer
  void unmap(c.MapType mtype, c.DimensionType dtype, ffi.Pointer<ffi.Void> mapPtr) =>
      c.mnn_tensor_unmap(ptr, mtype, dtype, mapPtr);

  /// @brief Wait for tensor ready
  ///
  /// @param mtype Map type
  ///
  /// @param finish Whether wait for finish
  ///
  /// @return Error code
  void wait(c.MapType mtype, bool finish) => c.mnn_tensor_wait(ptr, mtype, finish);

  /// @brief Set device pointer
  ///
  /// @param device_ptr Device pointer
  ///
  /// @param memory_type Memory type
  ///
  /// @return Error code
  void setDevicePtr(ffi.Pointer<ffi.Void> ptr, int memoryType) =>
      c.mnn_tensor_set_device_ptr(this.ptr, ptr, memoryType);

  void print() => c.mnn_tensor_print(ptr);

  void printShape() => c.mnn_tensor_print_shape(ptr);

  @override
  List<Object?> get props => [ptr.address];

  @override
  String toString() {
    return 'Tensor(address=0x${ptr.address.toRadixString(16)}, shape=$shape, type=$type, dimType=$dimensionType)';
  }
}

enum DataType {
  DataType_DT_INVALID(0),
  DataType_DT_FLOAT(1),
  DataType_DT_DOUBLE(2),
  DataType_DT_INT32(3),
  DataType_DT_UINT8(4),
  DataType_DT_INT16(5),
  DataType_DT_INT8(6),
  DataType_DT_STRING(7),
  DataType_DT_COMPLEX64(8),
  DataType_DT_INT64(9),
  DataType_DT_BOOL(10),
  DataType_DT_QINT8(11),
  DataType_DT_QUINT8(12),
  DataType_DT_QINT32(13),
  DataType_DT_BFLOAT16(14),
  DataType_DT_QINT16(15),
  DataType_DT_QUINT16(16),
  DataType_DT_UINT16(17),
  DataType_DT_COMPLEX128(18),
  DataType_DT_HALF(19),
  DataType_DT_RESOURCE(20),
  DataType_DT_VARIANT(21),
  DataType_MIN(0),
  DataType_MAX(21);

  final int value;
  const DataType(this.value);

  static DataType fromValue(int value) => switch (value) {
        0 => DataType_DT_INVALID,
        1 => DataType_DT_FLOAT,
        2 => DataType_DT_DOUBLE,
        3 => DataType_DT_INT32,
        4 => DataType_DT_UINT8,
        5 => DataType_DT_INT16,
        6 => DataType_DT_INT8,
        7 => DataType_DT_STRING,
        8 => DataType_DT_COMPLEX64,
        9 => DataType_DT_INT64,
        10 => DataType_DT_BOOL,
        11 => DataType_DT_QINT8,
        12 => DataType_DT_QUINT8,
        13 => DataType_DT_QINT32,
        14 => DataType_DT_BFLOAT16,
        15 => DataType_DT_QINT16,
        16 => DataType_DT_QUINT16,
        17 => DataType_DT_UINT16,
        18 => DataType_DT_COMPLEX128,
        19 => DataType_DT_HALF,
        20 => DataType_DT_RESOURCE,
        21 => DataType_DT_VARIANT,
        _ => throw Exception('Invalid value for DataType: $value'),
      };
}

extension DataTypeExt on DataType {
  /// @brief Convert DataType to HalideType
  ///
  /// @param type DataType
  ///
  /// @return HalideType
  HalideType toHalideType() {
    switch (this) {
      case DataType.DataType_DT_DOUBLE:
      case DataType.DataType_DT_FLOAT:
        return HalideType.f32();
      case DataType.DataType_DT_BFLOAT16:
        return HalideType.bf16();
      case DataType.DataType_DT_QINT32:
      case DataType.DataType_DT_INT32:
      case DataType.DataType_DT_BOOL:
      case DataType.DataType_DT_INT64:
        return HalideType.i32();
      case DataType.DataType_DT_QINT8:
      case DataType.DataType_DT_INT8:
        return HalideType.i8();
      case DataType.DataType_DT_QUINT8:
      case DataType.DataType_DT_UINT8:
        return HalideType.u8();
      case DataType.DataType_DT_QUINT16:
      case DataType.DataType_DT_UINT16:
        return HalideType.u16();
      case DataType.DataType_DT_QINT16:
      case DataType.DataType_DT_INT16:
        return HalideType.i16();
      default:
        throw Exception('Unsupported data type: $this');
    }
  }
}

File: src/tensor.cpp
/*
 * tensor.cpp
 * MNN C API for Tensor implementation
 *
 * This file implements the C-style API for MNN's Tensor functionality
 *
 * Author: Rainyl
 * License: Apache License 2.0
 */

#include "tensor.h"
#include "MNN/HalideRuntime.h"
#include "MNN/Tensor.hpp"
#include "error_code.h"
#include "mnn_type.h"
#include <cstdint>
#include <cstdlib>
#include <cstring>

enum DataType {
  DataType_DT_INVALID = 0,
  DataType_DT_FLOAT = 1,
  DataType_DT_DOUBLE = 2,
  DataType_DT_INT32 = 3,
  DataType_DT_UINT8 = 4,
  DataType_DT_INT16 = 5,
  DataType_DT_INT8 = 6,
  DataType_DT_STRING = 7,
  DataType_DT_COMPLEX64 = 8,
  DataType_DT_INT64 = 9,
  DataType_DT_BOOL = 10,
  DataType_DT_QINT8 = 11,
  DataType_DT_QUINT8 = 12,
  DataType_DT_QINT32 = 13,
  DataType_DT_BFLOAT16 = 14,
  DataType_DT_QINT16 = 15,
  DataType_DT_QUINT16 = 16,
  DataType_DT_UINT16 = 17,
  DataType_DT_COMPLEX128 = 18,
  DataType_DT_HALF = 19,
  DataType_DT_RESOURCE = 20,
  DataType_DT_VARIANT = 21,
  DataType_MIN = DataType_DT_INVALID,
  DataType_MAX = DataType_DT_VARIANT
};

halide_type_t mnn_halide_type_from_tensor_data_type(int type) {
  switch (type) {
  case DataType_DT_DOUBLE:
  case DataType_DT_FLOAT: return halide_type_of<float>(); break;
  case DataType_DT_BFLOAT16: return halide_type_t(halide_type_bfloat, 16); break;
  case DataType_DT_QINT32:
  case DataType_DT_INT32:
  case DataType_DT_BOOL:
  case DataType_DT_INT64: return halide_type_of<int32_t>(); break;
  case DataType_DT_QINT8:
  case DataType_DT_INT8: return halide_type_of<int8_t>(); break;
  case DataType_DT_QUINT8:
  case DataType_DT_UINT8: return halide_type_of<uint8_t>(); break;
  case DataType_DT_QUINT16:
  case DataType_DT_UINT16: return halide_type_of<uint16_t>(); break;
  case DataType_DT_QINT16:
  case DataType_DT_INT16: return halide_type_of<int16_t>(); break;
  default:
    MNN_PRINT("Unsupported data type! %d\n", type);
    MNN_ASSERT(false);
    return halide_type_of<float>(); // Return a default type for error handling
    break;
  }
}

int mnn_tensor_data_type_from_halide_type(halide_type_c_t type) {
  switch (type.code) {
  case halide_type_code_t::halide_type_uint:
    switch (type.bits) {
    case 8: return DataType_DT_UINT8; break;
    case 16: return DataType_DT_UINT16; break;
    default: return DataType_DT_INVALID; break;
    }
  case halide_type_code_t::halide_type_int:
    switch (type.bits) {
    case 8: return DataType_DT_INT8; break;
    case 16: return DataType_DT_INT16; break;
    case 32: return DataType_DT_INT32; break;
    case 64: return DataType_DT_INT64; break;
    default: return DataType_DT_INVALID; break;
    }
  case halide_type_code_t::halide_type_float:
    switch (type.bits) {
    case 32: return DataType_DT_FLOAT; break;
    case 64: return DataType_DT_DOUBLE; break;
    default: return DataType_DT_INVALID; break;
    }
  case halide_type_code_t::halide_type_bfloat:
    switch (type.bits) {
    case 16: return DataType_DT_BFLOAT16; break;
    default: return DataType_DT_INVALID; break;
    }
  default: return DataType_DT_INVALID; break;
  }
}

// Tensor creation functions
mnn_tensor_t mnn_tensor_create(int dim_size, mnn_dimension_type_t type) {
  try {
    MNN::Tensor *self = new MNN::Tensor(dim_size, (MNN::Tensor::DimensionType)type);
    return (mnn_tensor_t)self;
  } catch (...) { return nullptr; }
}

mnn_tensor_t
mnn_tensor_create_from_tensor(mnn_tensor_t self, mnn_dimension_type_t type, bool alloc_memory) {
  if (!self) return nullptr;
  try {
    MNN::Tensor *src = (MNN::Tensor *)self;
    MNN::Tensor *newTensor = new MNN::Tensor(src, (MNN::Tensor::DimensionType)type, alloc_memory);
    return (mnn_tensor_t)newTensor;
  } catch (...) { return nullptr; }
}

mnn_tensor_t mnn_tensor_create_device(
    const int *shape, int shape_size, halide_type_c_t type, mnn_dimension_type_t dim_type
) {
  if (!shape || shape_size <= 0) return nullptr;
  try {
    const auto _shape = std::vector<int>(shape, shape + shape_size);
    const auto _type = halide_type_t((halide_type_code_t)type.code, type.bits, type.lanes);
    MNN::Tensor *tensor =
        MNN::Tensor::createDevice(_shape, _type, (MNN::Tensor::DimensionType)dim_type);
    return (mnn_tensor_t)tensor;
  } catch (...) { return nullptr; }
}

mnn_tensor_t mnn_tensor_create_with_data(
    const int *shape,
    int shape_size,
    halide_type_c_t type,
    void *data,
    mnn_dimension_type_t dim_type
) {
  if (!shape || shape_size <= 0 || !data) return nullptr;
  try {
    auto _shape = std::vector<int>(shape, shape + shape_size);
    const auto _type = halide_type_t((halide_type_code_t)type.code, type.bits, type.lanes);
    MNN::Tensor *tensor =
        MNN::Tensor::create(_shape, _type, data, (MNN::Tensor::DimensionType)dim_type);
    return (mnn_tensor_t)tensor;
  } catch (...) { return nullptr; }
}

// MNN::Tensor destruction
void mnn_tensor_destroy(mnn_tensor_t self) {
  if (self) { MNN::Tensor::destroy((MNN::Tensor *)self); }
}

// MNN::Tensor operations
mnn_tensor_t mnn_tensor_clone(mnn_tensor_t src, bool deep_copy) {
  if (!src) return nullptr;
  try {
    MNN::Tensor *clone = MNN::Tensor::clone((MNN::Tensor *)src, deep_copy);
    return (mnn_tensor_t)clone;
  } catch (...) { return nullptr; }
}

mnn_error_code_t mnn_tensor_copy_from_host(mnn_tensor_t self, mnn_tensor_t host_tensor) {
  if (!self || !host_tensor) return INVALID_VALUE;
  try {
    MNN::Tensor *dst = (MNN::Tensor *)self;
    MNN::Tensor *src = (MNN::Tensor *)host_tensor;
    auto r = dst->copyFromHostTensor(src);
    return r ? BOOL_TRUE : BOOL_FALSE;
  } catch (...) { return UNKNOWN_ERROR; }
}

mnn_error_code_t mnn_tensor_copy_to_host(mnn_tensor_t self, mnn_tensor_t host_tensor) {
  if (!self || !host_tensor) return INVALID_VALUE;
  try {
    MNN::Tensor *src = (MNN::Tensor *)self;
    MNN::Tensor *dst = (MNN::Tensor *)host_tensor;
    auto r = src->copyToHostTensor(dst);
    return r ? BOOL_TRUE : BOOL_FALSE;
  } catch (...) { return UNKNOWN_ERROR; }
}

// MNN::Tensor properties
int mnn_tensor_dimensions(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->dimensions();
}

mnn_error_code_t mnn_tensor_shape(mnn_tensor_t self, int *shape, int shape_size) {
  if (!self || !shape) return INVALID_VALUE;
  MNN::Tensor *t = (MNN::Tensor *)self;
  if (shape_size < t->dimensions()) return INVALID_VALUE;

  for (int i = 0; i < t->dimensions(); i++) { shape[i] = t->length(i); }
  return NO_ERROR;
}

int mnn_tensor_size(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->size();
}

int mnn_tensor_element_size(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->elementSize();
}

size_t mnn_tensor_usize(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->usize();
}

// Dimension-specific properties
int mnn_tensor_width(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->width();
}

int mnn_tensor_height(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->height();
}

int mnn_tensor_channel(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->channel();
}

int mnn_tensor_batch(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->batch();
}

int mnn_tensor_stride(mnn_tensor_t self, int index) {
  if (!self || index < 0) return -1;
  return ((MNN::Tensor *)self)->stride(index);
}

int mnn_tensor_length(mnn_tensor_t self, int index) {
  if (!self || index < 0) return -1;
  return ((MNN::Tensor *)self)->length(index);
}

void mnn_tensor_set_stride(mnn_tensor_t self, int index, int stride) {
  if (self && index >= 0) { ((MNN::Tensor *)self)->setStride(index, stride); }
}

void mnn_tensor_set_length(mnn_tensor_t self, int index, int length) {
  if (self && index >= 0) { ((MNN::Tensor *)self)->setLength(index, length); }
}

// Data access
void *mnn_tensor_host(mnn_tensor_t self) {
  if (!self) return nullptr;
  return ((MNN::Tensor *)self)->host<void>();
}

uint64_t mnn_tensor_device_id(mnn_tensor_t self) {
  if (!self) return -1;
  return ((MNN::Tensor *)self)->deviceId();
}

halide_buffer_c_t *mnn_tensor_buffer(mnn_tensor_t self) {
  if (!self) return nullptr;
  auto _buf = ((MNN::Tensor *)self)->buffer();
  halide_buffer_c_t *buf = (halide_buffer_c_t *)malloc(sizeof(halide_buffer_c_t));
  buf->device = _buf.device;
  buf->device_interface = _buf.device_interface;
  buf->host = _buf.host;
  buf->flags = _buf.flags;
  buf->type = {(uint8_t)_buf.type.code, _buf.type.bits, _buf.type.lanes};
  buf->dimensions = _buf.dimensions;
  buf->dim = _buf.dim;
  buf->padding = _buf.padding;
  return buf;
}

// Type information
mnn_dimension_type_t mnn_tensor_get_dimension_type(mnn_tensor_t self) {
  if (!self) return MNN_TENSORFLOW;
  return (mnn_dimension_type_t)((MNN::Tensor *)self)->getDimensionType();
}

mnn_handle_data_type_t mnn_tensor_get_handle_data_type(mnn_tensor_t self) {
  if (!self) return MNN_HANDLE_NONE;
  return (mnn_handle_data_type_t)((MNN::Tensor *)self)->getHandleDataType();
}

void mnn_tensor_set_type(mnn_tensor_t self, int type) {
  if (self) { ((MNN::Tensor *)self)->setType(type); }
}

halide_type_c_t *mnn_tensor_get_type(mnn_tensor_t self) {
  if (!self) return nullptr;
  auto _type = ((MNN::Tensor *)self)->getType();
  halide_type_c_t *type = (halide_type_c_t *)malloc(sizeof(halide_type_c_t));
  type->code = (uint8_t)_type.code;
  type->bits = _type.bits;
  type->lanes = _type.lanes;
  return type;
}

// Memory mapping
void *mnn_tensor_map(mnn_tensor_t self, mnn_map_type_t mtype, mnn_dimension_type_t dtype) {
  if (!self) return nullptr;
  return ((MNN::Tensor *)self)->map((MNN::Tensor::MapType)mtype, (MNN::Tensor::DimensionType)dtype);
}

void mnn_tensor_unmap(
    mnn_tensor_t self, mnn_map_type_t mtype, mnn_dimension_type_t dtype, void *map_ptr
) {
  if (self && map_ptr) {
    ((MNN::Tensor *)self)
        ->unmap((MNN::Tensor::MapType)mtype, (MNN::Tensor::DimensionType)dtype, map_ptr);
  }
}

mnn_error_code_t mnn_tensor_wait(mnn_tensor_t self, mnn_map_type_t mtype, bool finish) {
  if (!self) return MNN_INVALID_PTR;
  try {
    ((MNN::Tensor *)self)->wait((MNN::Tensor::MapType)mtype, finish);
    return NO_ERROR;
  } catch (...) { return UNKNOWN_ERROR; }
}

mnn_error_code_t
mnn_tensor_set_device_ptr(mnn_tensor_t self, const void *device_ptr, int memory_type) {
  if (!self || !device_ptr) return MNN_INVALID_PTR;
  try {
    ((MNN::Tensor *)self)->setDevicePtr(device_ptr, memory_type);
    return NO_ERROR;
  } catch (...) { return UNKNOWN_ERROR; }
}

void mnn_tensor_print(mnn_tensor_t self) {
  if (!self) return;
  ((MNN::Tensor *)self)->print();
}

void mnn_tensor_print_shape(mnn_tensor_t self) {
  if (!self) return;
  ((MNN::Tensor *)self)->printShape();
}

MNN_C_API mnn_error_code_t mnn_tensor_set_image_f32(
    mnn_tensor_t self, int index, float *data, int width, int height, int channel
) {
  if (!self || !data) return MNN_INVALID_PTR;
  if (self->width() != width || self->height() != height || self->channel() != channel)
    return INVALID_VALUE;
  auto host = self->host<float>() + index * width * height * channel;
  memcpy(host, data, width * height * channel * sizeof(float));
  return NO_ERROR;
}
